sdk/src/account_context.rs

```diff
@@ -0,0 +1,258 @@
+        let instruction_context = self.current_invocation_frame()?;
+        instruction_context.borrow_account(
+            self,
+            instruction_context
+                .number_of_program_accounts
+                .saturating_sub(1),
+        )
+    }
+
+    pub fn borrow_instruction_account(
+        &self,
+        index_in_instruction: usize,
+    ) -> Result<BorrowedAccount, InstructionError> {
+        let instruction_context = self.current_invocation_frame()?;
+        instruction_context.borrow_account(
+            self,
+            instruction_context
+                .number_of_program_accounts
+                .saturating_add(index_in_instruction),
+        )
+    }
+
+    pub fn invocation_stack_push(
+        &mut self,
+        number_of_program_accounts: usize,
+        instruction_accounts: &[(usize, bool, bool)],
+        instruction_data: Vec<u8>,
+    ) -> Result<(), InstructionError> {
+        if self.invoke_stack.len() >= self.invoke_stack_max_height {
+            return Err(InstructionError::CallDepth);
+        }
+        let mut result = InstructionContext {
+            instruction_data,
+            number_of_program_accounts,
+            account_indices: Vec::with_capacity(instruction_accounts.len()),
+            account_is_signer: Vec::with_capacity(instruction_accounts.len()),
+            account_is_writable: Vec::with_capacity(instruction_accounts.len()),
+        };
+        for (index, is_signer, is_writable) in instruction_accounts.iter() {
+            result.account_indices.push(*index);
+            result.account_is_signer.push(*is_signer);
+            result.account_is_writable.push(*is_writable);
+        }
+        self.invoke_stack.push(result);
+        Ok(())
+    }
+
+    pub fn invocation_stack_pop(&mut self) -> Result<(), InstructionError> {
+        if self.invoke_stack.is_empty() {
+            return Err(InstructionError::CallDepth);
+        }
+        self.invoke_stack.pop();
+        Ok(())
+    }
+
+    pub fn set_return_data(&mut self, data: Vec<u8>) -> Result<(), InstructionError> {
+        let pubkey = *self.borrow_program_account()?.key();
+        self.return_data = (pubkey, data);
+        Ok(())
+    }
+
+    pub fn get_return_data(&self) -> (&Pubkey, &[u8]) {
+        (&self.return_data.0, &self.return_data.1)
+    }
+}
+
+/// Loaded instruction shared between runtime and programs.
+///
+/// This context is valid for the entire duration of a (possibly cross program) instruction being processed.
+pub struct InstructionContext {
+    number_of_program_accounts: usize,
+    account_indices: Vec<usize>,
+    account_is_signer: Vec<bool>,
+    account_is_writable: Vec<bool>,
+    instruction_data: Vec<u8>,
+}
+
+impl InstructionContext {
+    pub fn number_of_program_accounts(&self) -> usize {
+        self.number_of_program_accounts
+    }
+
+    pub fn number_of_accounts(&self) -> usize {
```

Comment:
Is this the total number of accounts or the number of non-program accounts?

Reply:
Added documentation and another method to differentiate that better